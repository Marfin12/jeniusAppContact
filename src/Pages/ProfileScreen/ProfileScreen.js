import * as React from 'react';
import { View } from 'react-native';
import { graphql, compose } from 'react-apollo';

import { DeleteContact, GetContact } from '../../Graphql/contact.graphql';
import AvatarImage from '../../Components/AvatarImage';
import FloatButton from '../../Components/FloatButton';
import Loading from '../../Components/Loading';
import TextField from '../../Components/TextField';

import styles from './ProfileScreen.styles';
import { deleteContactById, emptyContact } from './ProfileScreen.utils';

const ProfileScreen = (props) => {
  const [isLoading, setIsLoading] = React.useState(true);
  const [contact, setContact] = React.useState(emptyContact);
  const [firstNameText, setFirstName] = React.useState("");
  const [lastNameText, setLastName] = React.useState("");
  const [ageText, setAge] = React.useState("");
  const [photoText, setPhoto] = React.useState("");
  const [idText, setId] = React.useState("");

  React.useEffect(() => {
    const {firstName, lastName, age, photo, id} =
      props.data.getContactById ? props.data.getContactById : emptyContact; 
    setContact({
      firstName,
      lastName,
      age,
      photo,
      id
    })
    props.data.getContactById && setIsLoading(false);
  }, [props.data.getContactById]);

  React.useEffect(() => {
    const unsubscribe = props.navigation.addListener('focus', () => {
      setIsLoading(true);
      setTimeout(() => {
        setIsLoading(false);
      }, 2000)
      props.data.refetch();
      const {firstName, lastName, age, photo, id} =
        props.data.getContactById ? props.data.getContactById : emptyContact;
      setContact({
        firstName,
        lastName,
        age,
        photo,
        id
      })
    });
    
    return unsubscribe;
  }, [props.navigation]);

  React.useEffect(() => {
    setFirstName(contact.firstName);
    setLastName(contact.lastName);
    setAge(contact.age);
    setPhoto(contact.photo);
    setId(contact.id)
  }, [contact]);

  return (
    <View style={styles.container}>
      <Loading isVisible={isLoading} />
      <View style={styles.contentContainer}>
        <AvatarImage size={240} image={photoText} firstName={firstNameText} lastName={lastNameText} />
        <TextField style={styles.title} textStyle={styles.firstName}>
          {firstNameText}
        </TextField>
        <TextField style={styles.title} textStyle={styles.lastName}>
          {lastNameText}
        </TextField>
        <TextField style={styles.title} textStyle={styles.age}>
          {ageText}
        </TextField>
      </View>
      <FloatButton onPress={deleteContactById(idText, setIsLoading, props)} deleteButton />
    </View>
  );
};

export const queryOptions = (props: Props): QueryOpts => {
  return ({
  variables: {
    id: props.route.params.item.id
  }
})};

export const mapDeleteMutationToProps = ({ mutate }) => ({
  deleteContact: (id) =>
    mutate({ variables: { id }})
});

export default compose(
  graphql(DeleteContact, {
    props: mapDeleteMutationToProps
  }),
  graphql(GetContact, {
    options: queryOptions
  })
)(ProfileScreen);