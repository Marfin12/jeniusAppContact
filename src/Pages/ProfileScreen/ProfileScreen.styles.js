const styles = {
  container: {
    marginLeft: 10,
    justifyContent: 'center',
    alignItems: 'center',
  },
  description: {
    fontSize: 11,
    fontStyle: 'italic',
    marginTop: 10,
    width: '80%',
  },
  photo: {
    marginTop: 20,
    height: 200,
    width: '80%',
  },
  firstName: {
    fontSize: 24,
    fontWeight: 'bold',
    textAlign: 'center'
  },
  lastName: {
    fontSize: 16,
    textAlign: 'center',
    color: '#888888'
  },
  age: {
    fontSize: 16,
    textAlign: 'center',
  },
  title: {
    marginTop: 16,
  },
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: 12,
    paddingHorizontal: 32,
    borderRadius: 4,
    elevation: 3,
    backgroundColor: 'black',
  },
  buttonTitle: {
    fontSize: 16,
    lineHeight: 21,
    fontWeight: 'bold',
    letterSpacing: 0.25,
    color: 'white',
  },
  contentContainer: {
    marginTop: 20
  }
};

export default styles;
