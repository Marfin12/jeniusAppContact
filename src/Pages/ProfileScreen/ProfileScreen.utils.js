import { Alert } from 'react-native';

export const emptyContact = {
    firstName: "",
    lastName: "",
    age: "",
    id: "",
    photo: ""
};

const _isSuccessResponse = (message) => message === "contact deleted";

const _navigateToContactScreen = (isNavigation, navigation) => () => {
    if (isNavigation) navigation.goBack();
};

const _renderAlertDialog = (message, isNavigation, navigation) =>
  Alert.alert(
    _isSuccessResponse(message) ? "Success" : "Error",
    message,
    [
      { text: "OK", onPress: _navigateToContactScreen(isNavigation, navigation) }
    ]
  );

export const deleteContactById = (id, setIsLoading, props) => async () => {
  var message = "Something wrong occurs!!";
  var isNavigation = false;
  setIsLoading(true)
  try {
    const result = await props.deleteContact(id);
    setIsLoading(false);
    message = result.data.deleteContact.message;
    isNavigation = _isSuccessResponse(message);
  } catch (err) {
    setIsLoading(false);
    message = err.data.deleteContact.message
  } finally {
    _renderAlertDialog(message, isNavigation, props.navigation);
  }
};