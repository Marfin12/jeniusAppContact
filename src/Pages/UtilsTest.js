const apolloObjectCreator = (getListContact) => ({
    length: getListContact.length,
    getListContact,
  });
  
const apolloMapCreator = () => {
    const queryStore = new Map();
    queryStore.set({
      queryId: 123,
    });
  
    return queryStore;
};

export { apolloObjectCreator, apolloMapCreator };