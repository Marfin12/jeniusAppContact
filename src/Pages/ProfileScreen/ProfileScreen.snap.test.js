import React from 'react';
import renderer from 'react-test-renderer';
import { ApolloProvider } from 'react-apollo';

import { apolloObjectCreator, apolloMapCreator } from '../UtilsTest';
import ProfileScreen from './ProfileScreen';

jest.mock('../../Components/TextField', () => 'TextField');
jest.mock('../../Components/Card', () => 'Card');
jest.mock('../../Components/FloatButton', () => 'FloatButton');
jest.mock('../../Components/Loading', () => 'Loading');
jest.mock('../../Components/HeaderList', () => 'HeaderList');

const navigation = {
  addListener: jest.fn()
};
const client = {
  watchQuery: jest.fn(),
  refetch: false,
};
const mockedApolloConfig = {
  refetch: jest.fn(),
  fetchMore: jest.fn(),
  updateQuery: jest.fn(),
  startPolling: jest.fn(),
  stopPolling: jest.fn(),
  subscribeToMore: jest.fn(),
  options: {
    fetchPolicy: 'no-cache',
  },
  queryManager: {
    queryStore: apolloMapCreator(),
  },
  queryId: 123,
  subscribe: jest.fn(),
  currentResult: {
    loading: false,
    partial: true,
    networkStatus: 'ready',
    errors: null,
    error: null,
  },
};

let ProfileComponent;

afterEach(() => {
  jest.clearAllMocks();
});

describe('profile Screen Component with loading', () => {
  jest.mock('react-redux', () => {
    const ActualReactRedux = jest.requireActual('react-redux');
    const lightMode = {
      theme: false,
    };
    const darkMode = {
      theme: true,
    };
    return {
      ...ActualReactRedux,
      useSelector: jest
        .fn()
        .mockImplementationOnce(() => lightMode)
        .mockImplementationOnce(() => darkMode),
    };
  });

  let mockedProps = {
    route: {
      params: {
        item: {
          id: 1
        }
      }
    }
  }

  beforeEach(() => {
    client.watchQuery = jest.fn(() => ({
      ...mockedApolloConfig,
      currentResult: jest.fn().mockImplementation(() => ({
        ...mockedApolloConfig.currentResult,
        loading: true,
      })),
    }));
  });

  ProfileComponent = (
    <ApolloProvider client={client}>
      <ProfileScreen {...mockedProps} navigation={navigation} />
    </ApolloProvider>
  );

  test('renders light mode profile screen', () => {
    const ScreenWrapper = renderer.create(ProfileComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });

  test('renders dark mode profile screen', () => {
    const ScreenWrapper = renderer.create(ProfileComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });
});

describe('Profile Screen Component with failed to fetch', () => {
  jest.mock('react-redux', () => {
    const ActualReactRedux = jest.requireActual('react-redux');
    const lightMode = {
      theme: false,
    };
    const darkMode = {
      theme: true,
    };
    return {
      ...ActualReactRedux,
      useSelector: jest
        .fn()
        .mockImplementationOnce(() => lightMode)
        .mockImplementationOnce(() => darkMode),
    };
  });

  let mockedProps = {
    route: {
      params: {
        item: {
          id: 1
        }
      }
    }
  }

  beforeEach(() => {
    const getContactById = {};

    client.watchQuery = jest.fn(() => ({
      variables: getContactById,
      ...mockedApolloConfig,
      currentResult: jest.fn().mockImplementation(() => ({
        ...mockedApolloConfig.currentResult,
        data: apolloObjectCreator(getContactById),
      })),
    }));
  });

  ProfileComponent = (
    <ApolloProvider client={client}>
      <ProfileScreen {...mockedProps} navigation={navigation} />
    </ApolloProvider>
  );

  test('renders light mode profile screen', () => {
    const ScreenWrapper = renderer.create(ProfileComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });

  test('renders dark mode profile screen', () => {
    const ScreenWrapper = renderer.create(ProfileComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });
});

describe('Profile Screen Component with successfully fetch', () => {
  jest.mock('react-redux', () => {
    const ActualReactRedux = jest.requireActual('react-redux');
    const lightMode = {
      theme: false,
    };
    const darkMode = {
      theme: true,
    };
    return {
      ...ActualReactRedux,
      useSelector: jest
        .fn()
        .mockImplementationOnce(() => lightMode)
        .mockImplementationOnce(() => darkMode),
    };
  });

  beforeEach(() => {
    const getContactById = {
      firstName: 'Marfin',
      lastName: 'Fadhilah',
      age: 12,
      photo: 'example'
    };

    client.watchQuery = jest.fn(() => ({
      variables: getContactById,
      ...mockedApolloConfig,
      currentResult: jest.fn().mockImplementation(() => ({
        ...mockedApolloConfig.currentResult,
        data: apolloObjectCreator(getContactById),
      })),
    }));
  });

  let mockedProps = {
    route: {
      params: {
        item: {
          id: 1
        }
      }
    }
  }

  ProfileComponent = (
    <ApolloProvider client={client}>
      <ProfileScreen {...mockedProps} navigation={navigation} />
    </ApolloProvider>
  );

  test('renders light mode profile screen', () => {
    const ScreenWrapper = renderer.create(ProfileComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });

  test('renders dark mode profile screen', () => {
    const ScreenWrapper = renderer.create(ProfileComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });
});
