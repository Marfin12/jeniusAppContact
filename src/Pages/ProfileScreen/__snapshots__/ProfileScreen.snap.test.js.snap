// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen Component with failed to fetch renders dark mode home screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": 10,
    }
  }
>
  <Loading
    isVisible={true}
  />
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#f31222",
          "borderRadius": 120,
          "height": 240,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": 240,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "Inter-Medium",
            "fontSize": 120,
            "letterSpacing": 0,
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "color": "#888888",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
  </View>
  <FloatButton
    deleteButton={true}
    onPress={[Function]}
  />
</View>
`;

exports[`Home Screen Component with failed to fetch renders light mode home screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": 10,
    }
  }
>
  <Loading
    isVisible={true}
  />
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderRadius": 120,
          "height": 240,
          "marginLeft": 8,
          "overflow": "hidden",
          "width": 240,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "borderRadius": 120,
            "height": 240,
            "width": 240,
          }
        }
      />
    </View>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "color": "#888888",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
  </View>
  <FloatButton
    deleteButton={true}
    onPress={[Function]}
  />
</View>
`;

exports[`Home Screen Component with loading renders dark mode home screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": 10,
    }
  }
>
  <Loading
    isVisible={true}
  />
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#f31222",
          "borderRadius": 120,
          "height": 240,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": 240,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "Inter-Medium",
            "fontSize": 120,
            "letterSpacing": 0,
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "color": "#888888",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
  </View>
  <FloatButton
    deleteButton={true}
    onPress={[Function]}
  />
</View>
`;

exports[`Home Screen Component with loading renders light mode home screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": 10,
    }
  }
>
  <Loading
    isVisible={true}
  />
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderRadius": 120,
          "height": 240,
          "marginLeft": 8,
          "overflow": "hidden",
          "width": 240,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "borderRadius": 120,
            "height": 240,
            "width": 240,
          }
        }
      />
    </View>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "color": "#888888",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
  </View>
  <FloatButton
    deleteButton={true}
    onPress={[Function]}
  />
</View>
`;

exports[`Home Screen Component with successfully fetch renders dark mode home screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": 10,
    }
  }
>
  <Loading
    isVisible={true}
  />
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#f31222",
          "borderRadius": 120,
          "height": 240,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": 240,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "Inter-Medium",
            "fontSize": 120,
            "letterSpacing": 0,
            "textAlign": "center",
          }
        }
      >
        
      </Text>
    </View>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "color": "#888888",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
  </View>
  <FloatButton
    deleteButton={true}
    onPress={[Function]}
  />
</View>
`;

exports[`Home Screen Component with successfully fetch renders light mode home screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": 10,
    }
  }
>
  <Loading
    isVisible={true}
  />
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderRadius": 120,
          "height": 240,
          "marginLeft": 8,
          "overflow": "hidden",
          "width": 240,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "borderRadius": 120,
            "height": 240,
            "width": 240,
          }
        }
      />
    </View>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "color": "#888888",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
    <TextField
      style={
        Object {
          "marginTop": 16,
        }
      }
      textStyle={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      
    </TextField>
  </View>
  <FloatButton
    deleteButton={true}
    onPress={[Function]}
  />
</View>
`;
